src/runtime/gc_conservative.go

```diff
@@ -430,8 +430,18 @@ func markRoots(start, end uintptr) {
 		if start >= end {
 			runtimePanic("gc: unexpected range to mark")
 		}
+		if start%unsafe.Alignof(start) != 0 {
+			runtimePanic("gc: unaligned start pointer")
+		}
+		if end%unsafe.Alignof(end) != 0 {
+			runtimePanic("gc: unaligned end pointer")
+		}
 	}
 
+	// Reduce the end bound to avoid reading too far on platforms where pointer alignment is smaller than pointer size.
+	// If the size of the range is 0, then end will be slightly below start after this.
+	end -= unsafe.Sizeof(end) - unsafe.Alignof(end)
```

Comment:
:+1: nice, this should trivially compile down to nothing on non-AVR instruction sets.